### ISSUE TITLE
sr_disconnect race condition / lock timeout

### ISSUE THREAD

**Issue by gunterw** [sysrepo/sysrepo#2547]
My system has a bunch of processes with subscriptions and more processes constantly polling for operational data, everything works fine for a while and then a lock timeout (all different kinds) eventually occurs with some 'readers' never clearing out and things go south. After debugging this for a while I have figured out that the issue is something in the sr_disconnect sequence which causes a lock timeout in the same process that the sr_discconect was called. I was able to fix it by adding a flag in sr_conn_list_item and setting it to true in sr_connect and false at the begininng/first thing in sr_disconnect. I then use this new flag to determine if the CID is active in sr_shmmain_conn_check. I then added code to skip any "COND WAIT" logic if the owner CID is non-active. So basically stop waiting for any lock associated with the CID while it is being disconnected.

I'm currently using sysrepo version 1.4.136 (can't switch to libyang2 yet) but I've looked at the latest devel branch code and the places I have made changes are mostly the same.

I seem to be the only one with this problem, so I'm perfectly happy to keep my local mods and move on, but if there is interest in trying to get a fix into the devel branch let me know how to proceed.

**Comment by michalvasko** [comment]
We would definitely be interested in getting this fixed but you are experiencing this in v1 and even though the code that you have changed may be similar, locks are not and it seems the underlying problem is with some locks. So it would be best if you could reproduce it (in future) in *sysrepo* v2 and based on that we would proceed further.

**Comment by gunterw** [comment]
Ok, I will attempt to update my system to the latest devel branch commit to simplify debug and revisit this issue.

However, I am a bit confused as to where the line is drawn between sysrepo1 and sysrepo2

The latest legacy branch version is 0.7.9 (sysrepo1)
The master branch version I'm using is just before the latest devel->master merge 1.4.136 (?sysrepo1.5?)
The latest master version is 2.0.1 (sysrepo2)

The version I'm using is definitely not sysrepo1, but I guess it is also not really sysrepo2?

To be precise, the version I'm using is

commit d869c7ce7114b9437e6aa47d06ccc448e18bef0d
Author: Michal Vasko <mvasko@cesnet.cz>
Date:   Thu May 20 07:43:23 2021 +0200

    VERSION bump to version 1.4.136

and it definitely has all of the new sysrepo2 shared memory/locking stuff in it.

**Comment by michalvasko** [comment]
There is legacy *sysrepo* which should finally be put to rest and we do not use any specific version for that, I suppose it is version 0. Then there is version 1, which are simply all 1.x versions and version 2 that are all 2.x versions. As for the locks, most of the new stuff in v2 was ported over to v1 but I am pretty sure there are some improvements that are only in v2.

**Comment by gunterw** [comment]
ok, thanks for the clarification

**Comment by gunterw** [comment]
I keep re-reading this paragraph from the "Operational Subscriptions" documentation..

When using these pull operational data subscriptions, the subscription must be actively handling events whenever operational datastore data of the module are requested (so these 2 operations cannot occur in a single thread, so they must have individual subscription structures) to avoid a time out. 

and it is not clear to me what "2 operations" can't be in the same thread? 

Can you elaborate and/or give a complete example?

I'm asking because I'm using pull subscriptions for operational state data (not any config) and also using SR_SUBSCR_NO_THREAD for the subscriptions (with Glib::signal_io() for the event-pipe/process_events() callback), and I'm getting timeouts all the time (even with my mutex code changes mentioned above) so I'm wondering if I'm missing something.

**Comment by michalvasko** [comment]
The 2 operations are getting the operational data (such as using `sr_get_data()` or similar functions, is often called by other functions such as `sr_event_notif_send()`, `sr_rpc_send()` and similar) and providing the operational data (calling `sr_process_events()` on subscription that `sr_oper_get_items_subscribe()` was called on). And yes, it is more common if you use `SR_SUBSCR_NO_THREAD`.

**Comment by gunterw** [comment]
Sorry, still not getting it.. :-(.

Does this mean I can't call sr_get_data() from within the context of a subscription callback (process_events())?

If there is a requirement for multiple threads, it kinda breaks the whole reason that we use the Glib event loop/callback mechanism, we have everything run in one thread (the Glib event loop) so that we don't have to worry about threading issues,
everything is serialized.

**Comment by michalvasko** [comment]
> Does this mean I can't call sr_get_data() from within the context of a subscription callback (process_events())?

Yes, that should be correct.

> it kinda breaks the whole reason that we use the Glib event loop/callback mechanism

It would depend on the exact reason why `sr_get_data()` is being called but you can try some workarounds for this issue. Such as having the data provider run in a separate application.

**Comment by gunterw** [comment]
ok, this explains why I'm having all of these timeout/lockup issues!!
I think the answer for me is to convert all pull subscriptions to the push approach.
I assume this would eliminate this issue, correct?

**Comment by michalvasko** [comment]
Yes, that should also be possible.

**Comment by gunterw** [comment]
I assume you have considered using the PTHREAD_MUTEX_RECURSIVE attribute for mutexes?

Obviously they are a bit slower, but since the sysrepo library has little control over what happens in subscription callbacks,
it would be safer/more robust.

Maybe enable PTHREAD_MUTEX_RECURSIVE for SR_SUBSCR_NO_THREAD subscription mutexes by default,
since the issue is more likely to happen in this case?

Or a build time option to enable/disable PTHREAD_MUTEX_RECURSIVE for all mutexes and let the user make the tradeoff?

FYI - I have tried this and it does seem to help, but I'm not able to fully test things yet..

**Comment by michalvasko** [comment]
No, you have misunderstood the problem, recursive mutexes will not help here. The general architecture of subscriptions is based on events. So the issue is that whenever there is an event, it needs to be handled. And if you have only a single thread and call a function that generates an event handled by the same thread, you get a dead-lock and timeout because, obviously, that thread is not able to handle the event.

**Comment by gunterw** [comment]
fyi - just tried to update to the latest devel branch for sysrepo/netopeer2/libnetconf2/... to try and reproduce the issue above with the latest SW, but support for building deb packages for these projects appear to be missing, I assume support for this is on the list but not ready yet?

**Comment by michalvasko** [comment]
Yes, that is how it is.

**Comment by gunterw** [comment]
ok, having deb package support is a requirement for me (can only install deb packages in my embedded target environment), I'll update when this is available.

**Comment by jktjkt** [comment]
> Sorry, still not getting it.. :-(.
> 
> Does this mean I can't call sr_get_data() from within the context of a subscription callback (process_events())?
> 
> If there is a requirement for multiple threads, it kinda breaks the whole reason that we use the Glib event loop/callback mechanism, we have everything run in one thread (the Glib event loop) so that we don't have to worry about threading issues,
> everything is serialized.

Maybe I can help here. We, too, are using our own event loop, and we were hitting some deadlocks. In our case we were using the regular pull subscriptions for "on demand stuff", and we were also using a timer from the event loop to periodically push something to sysrepo. The problem was that when our timer fired and we pushed the data, sysrepo would try to validate the consistency of the operational DS, and that meant a callback to our pull-based callbacks. That one timed out because the event loop was still handling the timer event, so there was nobody to handle the FD-based event on a pull callback.

For us the fix was to either use push CBs everywhere, or pull CBs everywhere (and we ended up with push because wew anted to use YANG push for telemetry streaming anyway). Still on sysrepo 1.x (1.4 I think).

Hope this helps.

**Comment by gunterw** [comment]
Yep, this is what I'm doing now, converted everything to push CBs and all is well from that perspective. However I still think there is an issue with push CBs and config subscriptions happening at the same time that causes a lockup. Wanted to get to the latest devel 2.x branch to confirm and post info here, but currently stuck with 1.4 SW until deb packages are available. When I can upgrade to 2.x I will retry/retest everything and either post new info here or close this issue.
